lib/vquic/ngtcp2.c

```diff
@@ -833,18 +903,51 @@ static int cb_h3_recv_header(nghttp3_conn *conn, int64_t stream_id,
      !memcmp(":status", h3name.base, h3name.len)) {
     int status = decode_status_code(h3val.base, h3val.len);
     DEBUGASSERT(status != -1);
-    msnprintf(stream->mem, stream->len, "HTTP/3 %03d \r\n", status);
+    /* make sure there's enough room for the status line, which is 13 chars */
+    ncopy = 13;
+    if(stream->len >= ncopy) {
+      msnprintf(stream->mem, stream->len, "HTTP/3 %03d \r\n", status);
+      stream->len -= ncopy;
+      stream->memlen += ncopy;
+      stream->mem += ncopy;
+    }
+    else {
+      /* we have no space in the receive buffer, so use the overflow buffer */
+      result = allocate_overflow(data, stream, ncopy);
+      if(result) {
+        return -1;
+      }
+      char *buf = stream->overflow_buf + stream->overflow_buflen;
+      size_t size = stream->overflow_bufsize - stream->overflow_buflen;
+      msnprintf(buf, size, "HTTP/3 %03d \r\n", status);
+      stream->overflow_buflen += ncopy;
+    }
   }
   else {
     /* store as a HTTP1-style header */
-    msnprintf(stream->mem, stream->len, "%.*s: %.*s\n",
-              h3name.len, h3name.base, h3val.len, h3val.base);
+    ncopy = h3name.len + 2 + h3val.len + 1;
+    if(stream->len >= ncopy) {
+      msnprintf(stream->mem, stream->len, "%.*s: %.*s\n",
+                h3name.len, h3name.base, h3val.len, h3val.base);
+      stream->len -= ncopy;
+      stream->memlen += ncopy;
+      stream->mem += ncopy;
+    }
+    else {
+      /* we have no space in the receive buffer, so use the overflow buffer */
+      result = allocate_overflow(data, stream, ncopy);
+      if(result) {
+        return -1;
+      }
+      char *buf = stream->overflow_buf + stream->overflow_buflen;
+      size_t size = stream->overflow_bufsize - stream->overflow_buflen;
```

Comment:
The variables need to be declared at the top of the block - C89 compliance you know... =) Obviously the ngtcp2 build isn't building with the picky compiler options that otherwise would warn/error on this.

Reply:
Fixed.